app/email_service.py

```
@@ -0,0 +1,895 @@
+        self.rsi_threshold = 5.0  # RSI values within 5 points are considered similar
+        self.alert_cooldowns = {}  # {alert_hash: last_sent_timestamp}
+        self.alert_values = {}  # {alert_hash: last_sent_values} for value comparison
+        
+        if not self.sendgrid_api_key:
+            logger.warning("⚠️ SendGrid API key not found. Email notifications will be disabled.")
+            self.sg = None
+        else:
+            self.sg = SendGridAPIClient(api_key=self.sendgrid_api_key)
+            logger.info("✅ SendGrid email service initialized with smart value-based cooldown (10min, 5 RSI threshold)")
+    
+    def _generate_alert_hash(self, user_email: str, alert_name: str, triggered_pairs: List[Dict[str, Any]], calculation_mode: str = None) -> str:
```

Ruff (0.12.2)

PEP 484 prohibits implicit `Optional`

Convert to `Optional[T]`

(RUF013)